sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,461 @@
+
+[source,c++]
+----
+
+namespace sycl {
+
+event submit(const ext::oneapi::experimental::graph<ext::oneapi::experimental::graph_state::executable>& my_graph);
+
+} // namespace sycl
+
+----
+
+=== Executable Graph
+
+A `graph` object in `graph_state::executable` represents a user generated device and context specific execution object that is submitted to a queue for execution.
+The structure of such a `graph` object in this state is immutable and cannot be changed, so are the tasks assigned with each node.
+Support of submitting a graph for execution, before a previous execution has been completed is backend specific. The runtime may throw an error.
+
+=== Graph member and helper functions
+
+Table 3. Constructor of the `graph` class.
+[cols="2a,a"]
+|===
+|Constructor|Description
+
+|
+[source,c++]
+----
+/* available only when graph_state == modifiable */`
+graph();
+----
+|Creates a `graph` object.
+
+|===
+
+Table 4. Member functions of the `graph` class.
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+node add_node(const std::vector<node>& dep = {});
+----
+|This creates an empty node which is associated to no task. Its intended use is either a connection point inside a graph between groups of nodes, and can significantly reduce the number of edges ( O(n) vs. O(n^2) ). Another use-case is building the structure of a graph first and adding tasks later.
+
+|
+[source,c++]
+----
+template<typename T>
+    node add_node(T cgf, const std::vector<node>& dep = {});
+----
+|This node captures a command group function object containing host task which is scheduled by the SYCL runtime or a SYCL function for invoking kernels with all restrictions that apply as described in the spec.
+
+|===
+
+Memory that is allocated by the following functions is owned by the specific graph. When freed inside the graph, the memory is only accessible before the `free` node is executed and after the `malloc` node is executed.
+
+Table 5. Member functions of the `graph` class (memory operations).
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+node memcpy(void* dest, const void* src, size_t numBytes, const std::vector<node>& dep = {});
+----
+|Adding a node that encapsulates a `memcpy` operation.
+
+|
+[source,c++]
+----
+template<typename T> node
+copy(const T* src, T* dest, size_t count, const std::vector<node>& dep = {});
+----
+|Adding a node that encapsulates a `copy` operation.
+
+|
+[source,c++]
+----
+node memset(void* ptr, int value, size_t numBytes, const std::vector<node>& dep = {});
+----
+|Adding a node that encapsulates a `memset` operation.
+
+|
+[source,c++]
+----
+template<typename T>
+node fill(void* ptr, const T& pattern, size_t count, const std::vector<node>& dep = {});
+----
+|Adding a node that encapsulates a `fill` operation.
+
+|
+[source,c++]
+----
+node malloc(void *data, size_t numBytes, usm::alloc kind, const std::vector<node>& dep = {});
```

Comment:
After looking at your code:
```suggestion
node malloc(void*& data, size_t numBytes, usm::alloc kind, const std::vector<node>& dep = {});
```
but it looks like a C interface and and ugly C++.

Reply:
Good point and agree this needs improvement.